exporter/awsprometheusremotewriteexporter/auth.go

```diff
@@ -0,0 +1,99 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsprometheusremotewriteexporter
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+)
+
+// signingRoundTripper is a Custom RoundTripper that performs AWS Sig V4
+type signingRoundTripper struct {
+	transport http.RoundTripper
+	signer    *v4.Signer
+	cfg       *aws.Config
+	service   string
+}
+
+// RoundTrip signs each outgoing request
+func (si *signingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
```

Comment:
A note about the RoundTripper implementation. The RoundTripper contract is very peculiar about not mutating the incoming request object. That's why cloning the request and incoming body is preferred. 

See from the godoc,

"A RoundTripper must be safe for concurrent use by multiple goroutines." 
https://golang.org/pkg/net/http/#RoundTripper

When a RoundTripper modifies the request and the body, it becomes unsafe for concurrent use unless you introduce locks.

I don't know who's going to call this roundtripper and whether the same req can ever be passed. If that's the case, cloning the request and the body might be an option. This is what we did earlier for the request object (we didn't slurp the body in the roundtripper so it's untouched): https://github.com/golang/oauth2/blob/master/transport.go#L79

Reply:
This http client is only used synchronously within the Prometheus Remote Write Exporter so it should not be a problem however I will make copies to conform with how Go expects RoundTrippers to be implemented